### ISSUE TITLE
prop-types validation should resolve through object destruct of this.props

### ISSUE THREAD

**Issue by cassus** [yannickcr/eslint-plugin-react#162]
```js
export class Application {
  static propTypes = {
    user: PropTypes.shape({
      displayName: PropTypes.string,
    }),
  }
  render() {
    const { user } = this.props
    return <div className="ApplicationComponent">
         {user.username} {/* <---- should warn */}
      </div>
      }
}
```

**Comment by cassus** [comment]
I see this style spreading, it was in some presentations on React Europe conf. It's also the recommended style in https://github.com/gaearon/redux

**Comment by yannickcr** [comment]
This is not planned for now. This would require a lot of additional analysis (and likely an additional static analysis tool) to find out if a variable is referring to a prop at some point.

For now we only support this form of destructuring (much easier to analyse):

```js
export class Application {
  static propTypes = {
    user: PropTypes.shape({
      displayName: PropTypes.string,
    }),
  }
  render() {
    const { username } = this.props.user /* <--- warn */
    return <div className="ApplicationComponent">
         {username}
      </div>
      }
}
```

**Comment by cassus** [comment]
Good workaround, thank you!

Shall we leave this issue open?

**Comment by yannickcr** [comment]
If you consider this issue resolved you can close it.
